### PR TITLE
SIMDify interleaving the RGB(A) pixels for the decoder callback.

### PR DESCRIPTION
**Pull request by szabadka**

### CODE REVIEW

File: lib/jxl/render_pipeline/stage_write.cc [hunk 1]
```diff
@@ -356,20 +322,193 @@ class WriteToPixelCallbackStage : public RenderPipelineStage {
...
+  void StoreUnsignedRow(const float* input[4], size_t len, T* output) const {
+    const HWY_FULL(float) d;
+    auto zero = Zero(d);
+    auto one = Set(d, 1.0f);
+    auto mul = Set(d, (1u << (sizeof(T) * 8)) - 1);
+    const Rebind<T, decltype(d)> du;
+    for (size_t c = 0; c < num_channels_; ++c) {
+      UnpoisonPadding(d, len, input[c]);
+    }
+    UnpoisonPadding(d, len * num_channels_, output);
```
Comment (veluca93) [inline lib/jxl/render_pipeline/stage_write.cc:384]: Why are we doing this? I assume we only store to the output, no?

File: lib/jxl/render_pipeline/stage_write.cc [hunk 2]
```diff
@@ -356,20 +322,193 @@ class WriteToPixelCallbackStage : public RenderPipelineStage {
...
+        auto v0 = Mul(Clamp(zero, LoadU(d, &input[0][i]), one), mul);
+        auto v1 = Mul(Clamp(zero, LoadU(d, &input[1][i]), one), mul);
+        auto v2 = Mul(Clamp(zero, LoadU(d, &input[2][i]), one), mul);
+        auto v3 = Mul(Clamp(zero, LoadU(d, &input[3][i]), one), mul);
+        StoreInterleaved4(DemoteTo(du, NearestInt(v0)),
+                          DemoteTo(du, NearestInt(v1)),
+                          DemoteTo(du, NearestInt(v2)),
+                          DemoteTo(du, NearestInt(v3)), du, &output[4 * i]);
+      }
+    }
```
Comment (veluca93) [inline lib/jxl/render_pipeline/stage_write.cc:417]: We should re-poison the output padding here

File: lib/jxl/render_pipeline/stage_write.cc [hunk 3]
```diff
@@ -356,20 +322,193 @@ class WriteToPixelCallbackStage : public RenderPipelineStage {
...
+    for (size_t c = 0; c < num_channels_; ++c) {
+      line_buffers[c] = temp_in[c];
+    }
+    *xstart = width_ - *xstart - len;
+  }
+
   template <typename T>
-  void WriteToCallback(size_t thread_id, size_t ypos, size_t xstart,
-                       size_t xlen, T* output) const {
+  void StoreUnsignedRow(const float* input[4], size_t len, T* output) const {
```
Comment (veluca93) [inline lib/jxl/render_pipeline/stage_write.cc:375]: Does this work with 16 or 8 bit types?

File: lib/jxl/render_pipeline/stage_write.cc [hunk 4]
```diff
@@ -356,20 +322,193 @@ class WriteToPixelCallbackStage : public RenderPipelineStage {
...
+
+  void StoreFloat16Row(const float* input[4], size_t len,
+                       uint16_t* output) const {
+    const HWY_FULL(float) d;
+    const Rebind<uint16_t, decltype(d)> du;
+    const Rebind<hwy::float16_t, decltype(d)> df16;
+    for (size_t c = 0; c < num_channels_; ++c) {
+      UnpoisonPadding(d, len, input[c]);
+    }
+    UnpoisonPadding(d, len * num_channels_, output);
```
Comment (veluca93) [inline lib/jxl/render_pipeline/stage_write.cc:428]: See comments above

**Review by veluca93 — APPROVED** [review]

File: lib/jxl/render_pipeline/stage_write.cc (hunk 1 above)
Comment (szabadka) [reply lib/jxl/render_pipeline/stage_write.cc]: Removed.

File: lib/jxl/render_pipeline/stage_write.cc (hunk 2 above)
Comment (szabadka) [reply lib/jxl/render_pipeline/stage_write.cc:414]: Removed the unpoisoning.

File: lib/jxl/render_pipeline/stage_write.cc (hunk 3 above)
Comment (szabadka) [reply lib/jxl/render_pipeline/stage_write.cc:373]: It works with uint8_t and uint16_t.

File: lib/jxl/render_pipeline/stage_write.cc (hunk 4 above)
Comment (szabadka) [reply lib/jxl/render_pipeline/stage_write.cc]: Removed.

File: lib/jxl/render_pipeline/stage_write.cc (hunk 3 above)
Comment (veluca93) [reply lib/jxl/render_pipeline/stage_write.cc:373]: Uh, OK, I thought DemoteTo didn't... Oh well, better this way :)

File: lib/jxl/render_pipeline/stage_write.cc (hunk 2 above)
Comment (veluca93) [reply lib/jxl/render_pipeline/stage_write.cc:414]: Even then: output values that came from the unpoisoned inputs should be poisoned now.

**Review by veluca93 — DISMISSED** [review]

File: lib/jxl/render_pipeline/stage_write.cc (hunk 2 above)
Comment (szabadka) [reply lib/jxl/render_pipeline/stage_write.cc:420]: Ah, I see, I poisoned the output in both places.

**Review by veluca93 — APPROVED** [review]

**Pull request merged**